### PR TITLE
Fixed #32437 -- Fixed cleaning up ALLOWED_HOSTS in LiveServerTestCase on setUpClass() failure.

### PR DESCRIPTION
**Pull request by cjerdonek**
https://code.djangoproject.com/ticket/32437

### CODE REVIEW

File: tests/servers/tests.py [hunk 1]
```diff
@@ -39,6 +40,45 @@ def urlopen(self, url):
...
+            raise RuntimeError(f'{settings.ALLOWED_HOSTS} != {expected}')
+
+    @classmethod
+    def setUpClass(cls):
+        cls.checkAllowedHosts(['testserver'])
+        try:
+            super().setUpClass()
+        except RuntimeError:
+            # Check that LiveServerTestCase's change to ALLOWED_HOSTS
+            # was undone.
```
Comment (felixxm) [inline tests/servers/tests.py:63]:
```suggestion
            # LiveServerTestCase's change to ALLOWED_HOSTS should be reverted.
```

File: tests/servers/tests.py [hunk 2]
```diff
@@ -39,6 +40,45 @@ def urlopen(self, url):
...
+class FailingLiveServerThread(LiveServerThread):
+    def _create_server(self):
+        raise RuntimeError('failure creating server')
+
+
+class LiveServerTestCaseSetupTest(LiveServerBase):
+    server_thread_class = FailingLiveServerThread
+
+    @classmethod
+    def checkAllowedHosts(cls, expected):
```
Comment (felixxm) [inline tests/servers/tests.py:52]:
```suggestion
    def check_allowed_hosts(cls, expected):
```

File: tests/servers/tests.py [hunk 3]
```diff
@@ -39,6 +40,45 @@ def urlopen(self, url):
...
+        try:
+            super().setUpClass()
+        except RuntimeError:
+            # Check that LiveServerTestCase's change to ALLOWED_HOSTS
+            # was undone.
+            cls.checkAllowedHosts(['testserver'])
+        else:
+            raise RuntimeError('server did not fail')
+
+        cls._setUpCalled = True
```
Comment (felixxm) [inline tests/servers/tests.py:68]:
```suggestion
        cls.set_up_called = True
```

File: tests/servers/tests.py [hunk 4]
```diff
@@ -39,6 +40,45 @@ def urlopen(self, url):
...
+    @classmethod
+    def tearDownClass(cls):
+        # Make tearDownClass() a no-op because setUpClass() was already
+        # cleaned up, and because the error inside setUpClass() was handled,
+        # which will cause tearDownClass() to be called when it normally
+        # wouldn't.
+        pass
+
+    def test_set_up_class(self):
+        assert self._setUpCalled
```
Comment (felixxm) [inline tests/servers/tests.py:79]:
```suggestion
        assertIs(self.set_up_called, True)
```

**Review by felixxm — COMMENTED** [review]
@cjerdonek Thanks for this patch :+1: I left some minor stylistic comments.

File: tests/servers/tests.py (hunk 4 above)
Comment (cjerdonek) [reply tests/servers/tests.py:79]: `self.assertIs(...`

**Comment by cjerdonek** [comment]
Thanks, @felixxm! I incorporated your suggestions in a new commit.

**Review by apollo13 — APPROVED** [review]
Changes look good to me; that said I fully expected that commit to come back and bite me -- see the commit msg in https://github.com/django/django/commit/73a610d2a81bc3bf2d3834786b2458bc85953ed0 :D

**Comment by cjerdonek** [comment]
Thanks, @apollo13. :) (FWIW, I did see that commit and referenced it in a related issue here: https://code.djangoproject.com/ticket/32417#comment:7 )

**Comment by felixxm** [comment]
Thanks both :+1:

**Pull request merged**